net/isoping/isoping.go

```diff
@@ -0,0 +1,203 @@
+	usecPerPrint   int32
+	nextTxId       uint32
+	nextRxId       uint32
+	nextRxackId    uint32
+	startRtxtime   uint32 // remote's txtime at startup
+	startRxtime    uint32 // local rxtime at startup
+	lastRxtime     uint32 // local rxtime of last received packet
+	minCycleRxdiff int32  // smallest packet delay seen this cycle
+	nextCycle      uint32 // time when next cycle begins
+	now            uint32 // current time
+	nextSend       uint32 // time when we'll send next pkt
+	numLost        uint32 // number of rx packets not received
+	nextTxackIndex int    // next array item to fill in tx.acks
+	lastPrint      uint32 // time of last packet printout
+	latTx          int64
+	latTxMin       int64
+	latTxMax       int64
+	latTxCount     int64
+	latTxSum       int64
+	latTxVarSum    int64
+
+	latRx       int64
+	latRxMin    int64
+	latRxMax    int64
+	latRxCount  int64
+	latRxSum    int64
+	latRxVarSum int64
+}
+
+// Incremental standard deviation calculation, without needing to know the
+// mean in advance.  See:
+// http://mathcentral.uregina.ca/QQ/database/QQ.09.02/carlos1.html
+func onepass_stddev(sumsq int64, sum int64, count int64) float64 {
+	numer := (count * sumsq) - (sum * sum)
+	denom := count * (count - 1)
+	return math.Sqrt(DIV(numer, denom))
+}
+
+// ustimenow subtracts the time since the program started and returns it
+func (srv *Isoping) ustimenow() uint64 {
+	tn := time.Since(srv.ClockStartTime)
+	return uint64(tn.Microseconds())
+}
+
+// Ustime casts the result of ustimenow to uint32 and returns it
+func (srv *Isoping) Ustime() uint32 {
+	return uint32(srv.ustimenow())
+}
+
+// initClock keeps track of when the server/client starts.
+// keeps the exact time and we can subtract from the time
+// to get monotonicClock values
+func (srv *Isoping) initClock() {
+	srv.ClockStartTime = time.Now()
+}
+
+// initClient sets the Isoping.Conn, to the address string otherwise
+// uses [::]:4948 as the default
+func (srv *Isoping) initClient(addressString string) {
+	srv.initClock()
+	srv.IsServer = false
+	udpaddr, err := net.ResolveUDPAddr("udp6", addressString)
+	if err != nil {
+		log.Println(err)
+		addr := "[::]" + SERVER_PORT
+		udpaddr, err = net.ResolveUDPAddr("udp6", addr)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+		log.Printf("Address %v failed to resolve, using %v instead\n", addressString, udpaddr)
+	}
+
+	conn, err := net.DialUDP("udp6", nil, udpaddr)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	srv.RemoteAddr = udpaddr
+	srv.Conn = conn
+}
+
+// initServer sets the Conn field of Isoping, for the listener side.
+func (srv *Isoping) initServer() {
+	srv.initClock()
+	srv.IsServer = true
+	addr, err := net.ResolveUDPAddr("udp6", SERVER_PORT)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	srv.ListenAddr = addr
+	srv.Conn, err = net.ListenUDP("udp", addr)
+	if err != nil {
+		log.Printf("%v\n", err)
+		return
+	}
+}
+
+// initVars initializes a lot of the necessary variables for the calculation
```

Comment:
This is a case where I think the Go code should diverge more from C. C only added named struct initializers in C99:
```
struct Isoping{.nextTxId=1, .nextRxId=0, ...
```

For maximal portability with older C, important back when this C code was originally written, you couldn't use named initializers and added functions like this instead.

Since initVars is called from only one place, I'd recommend instead putting the field initialization into the struct when it is created.

Reply:
I will make a constructor function for this!